controllers/kserve_ray_tls_controller.go

```diff
@@ -0,0 +1,404 @@
+package controllers
+
+import (
+	"context"
+	"os"
+	"reflect"
+
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+
+	"github.com/go-logr/logr"
+	kservev1alpha1 "github.com/kserve/kserve/pkg/apis/serving/v1alpha1"
+	"github.com/opendatahub-io/odh-model-controller/controllers/constants"
+	"github.com/opendatahub-io/odh-model-controller/controllers/utils"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+)
+
+// KServeRayTlsReconciler holds the controller configuration.
+type KServeRayTlsReconciler struct {
+	client client.Client
+	log    logr.Logger
+}
+
+func NewKServeRayTlsReconciler(client client.Client, log logr.Logger) *KServeRayTlsReconciler {
+	return &KServeRayTlsReconciler{
+		client: client,
+		log:    log,
+	}
+}
+
+// The reconcile logic works as follows:
+// ServingRuntime(multinode):
+// - On creation: The ray-tls-script ConfigMap and ray-ca-cert Secret are created in the target namespace.
+// - On deletion: The ray-tls-script ConfigMap and ray-ca-cert Secret are deleted only when multinode ServingRuntimes are deleted from the target namespace.
+
+// ConfigMap:
+// - When the original ConfigMap is updated in the control namespace: The ray-tls-scripts ConfigMap is deleted and recreated in the namespace where multinode ServingRuntimes exist.
+// - When the ConfigMap is deleted in the target namespace: The ray-tls-scripts ConfigMap will be recreated.
+
+// Secret:
+// - When the original Secret is updated in the control namespace: The ray-ca-cert Secret is deleted and recreated in the namespace where multinode ServingRuntimes exist.
+// - When the Secret is deleted in the target namespace: The ray-ca-cert Secret will be recreated.
+func (r *KServeRayTlsReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.log
+	controllerNs := os.Getenv("POD_NAMESPACE")
+	var servingRuntimeList kservev1alpha1.ServingRuntimeList
+	if err := r.client.List(ctx, &servingRuntimeList); err != nil {
+		return ctrl.Result{}, err
+	}
+	noMultiNodeSrExistInNs := !existMultiNodeServingRuntimeInNs(req.Namespace, servingRuntimeList)
+
+	if req.Name == constants.RayTlsScriptConfigMapName {
+		if req.Namespace == controllerNs {
+			log.Info("Original Ray TLS Scripts ConfigMap is updated", "name", constants.RayTlsScriptConfigMapName, "namespace", req.Namespace)
```

Comment:
as user will not configure it at this time, maybe messages like this could be debug, wdyt?

Reply:
for upgrading time, I wanted to see this message if we changes this files. I should not be shown up in usual case unless somebody changes it.